sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeArrayWriter.java

```diff
@@ -33,91 +38,147 @@
   // The offset of the global buffer where we start to write this array.
   private int startingOffset;
 
+  // The number of elements in this array
+  private int numElements;
+
+  private int headerInBytes;
+
+  private void assertIndexIsValid(int index) {
+    assert index >= 0 : "index (" + index + ") should >= 0";
+    assert index < numElements : "index (" + index + ") should < " + numElements;
+  }
+
   public void initialize(BufferHolder holder, int numElements, int fixedElementSize) {
-    // We need 4 bytes to store numElements and 4 bytes each element to store offset.
```

Comment:
We still need to comment about the `numElements` stuff, or it will confuse other reader when they see `startingOffset + 4` later.